# Real-world Data Wrangling

## Data No 1: 120 years of Olympic history: athletes and results
## Data No 2: Wrangled Pakistan largest ecommerce dataset

### 1. Gather data

#### Data No 1:
120 years of olympic history is downloaded using kaggle api. This involves installing pip kaggle. Secondly, generating kaggle token. Then, we copied the dataset api and we download data using kaggle api. Lastly, we unzip this data by importing zipfile and then we assess data using pandas.

#### Data No 2:
Pakistan largest ecommerce dataset is downloaded manually from kaggle and start using with pandas.

# **Dataset 1:**
### Gathering data by accessing kaggle API.


In [1]:
#Importing all useful directories for this project:

import pandas as pd
import numpy as np
import zipfile
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns 

%matplotlib inline

In [2]:
# Installing kaggle
!pip install kaggle

In [3]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
! chmod 600 ~/.kaggle/kaggle.json

In [4]:
#Downloading dataset kaggle api
!kaggle datasets download -d heesoo37/120-years-of-olympic-history-athletes-and-results

120-years-of-olympic-history-athletes-and-results.zip: Skipping, found more recently modified local copy (use --force to force download)


In [5]:
dataset = '/Users/bilal/Downloads/Project/120-years-of-olympic-history-athletes-and-results.zip'

In [6]:
# Extracting the files from the downloaded zipfile
with zipfile.ZipFile(dataset, 'r') as zip_ref:
    zip_ref.extractall()

In [7]:
df_olympic = pd.read_csv('athlete_events.csv')
df_olympic.head(10)

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   
5   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   
6   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0     Netherlands   
7   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0     Netherlands   
8   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0     Netherlands   
9   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season         City          Sport  \
0  CHN  1992 Summer  1992  Summer    Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer       London           Judo   
2  DEN  1920 Summer  1920  Summer    Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer        Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter      Calgary  Speed Skating   
5  NED  1988 Winter  1988  Winter      Calgary  Speed Skating   
6  NED  1992 Winter  1992  Winter  Albertville  Speed Skating   
7  NED  1992 Winter  1992  Winter  Albertville  Speed Skating   
8  NED  1994 Winter  1994  Winter  Lillehammer  Speed Skating   
9  NED  1994 Winter  1994  Winter  Lillehammer  Speed Skating   

                                Event Medal  
0         Basketball Men's Basketball   NaN  
1        Judo Men's Extra-Lightweight   NaN  
2             Football Men's Football   NaN  
3         Tug-Of-War Men's Tug-Of-War  Gold  
4    Speed Skating Women's 500 metres   NaN  
5  Speed Skating Women's 1,000 metres   NaN  
6    Speed Skating Women's 500 metres   NaN  
7  Speed Skating Women's 1,000 metres   NaN  
8    Speed Skating Women's 500 metres   NaN  
9  Speed Skating Women's 1,000 metres   NaN

# Dataset 2
### Download data manually from kaggle.

In [8]:
data_environment = pd.read_csv('Environment_Temperature_change_E_All_Data_NOFLAG.csv',encoding='latin-1')
data_environment.head(5)


Area Code         Area  Months Code    Months  Element Code  \
0          2  Afghanistan         7001   January          7271   
1          2  Afghanistan         7001   January          6078   
2          2  Afghanistan         7002  February          7271   
3          2  Afghanistan         7002  February          6078   
4          2  Afghanistan         7003     March          7271   

              Element Unit  Y1961  Y1962  Y1963  ...  Y2010  Y2011  Y2012  \
0  Temperature change   °C  0.777  0.062  2.744  ...  3.601  1.179 -0.583   
1  Standard Deviation   °C  1.950  1.950  1.950  ...  1.950  1.950  1.950   
2  Temperature change   °C -1.743  2.465  3.919  ...  1.212  0.321 -3.201   
3  Standard Deviation   °C  2.597  2.597  2.597  ...  2.597  2.597  2.597   
4  Temperature change   °C  0.516  1.336  0.403  ...  3.390  0.748 -0.527   

   Y2013  Y2014  Y2015  Y2016  Y2017  Y2018  Y2019  
0  1.233  1.755  1.943  3.416  1.201  1.996  2.951  
1  1.950  1.950  1.950  1.950  1.950  1.950  1.950  
2  1.494 -3.187  2.699  2.251 -0.323  2.705  0.086  
3  2.597  2.597  2.597  2.597  2.597  2.597  2.597  
4  2.246 -0.076 -0.497  2.296  0.834  4.418  0.234  

[5 rows x 66 columns]

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**

### Quality Issue 1:

In [9]:
df_olympic.describe()
df_olympic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271116 entries, 0 to 271115
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   ID      271116 non-null  int64  
 1   Name    271116 non-null  object 
 2   Sex     271116 non-null  object 
 3   Age     261642 non-null  float64
 4   Height  210945 non-null  float64
 5   Weight  208241 non-null  float64
 6   Team    271116 non-null  object 
 7   NOC     271116 non-null  object 
 8   Games   271116 non-null  object 
 9   Year    271116 non-null  int64  
 10  Season  271116 non-null  object 
 11  City    271116 non-null  object 
 12  Sport   271116 non-null  object 
 13  Event   271116 non-null  object 
 14  Medal   39783 non-null   object 
dtypes: float64(3), int64(2), object(10)
memory usage: 31.0+ MB


In [10]:
#FILL IN - Inspecting the dataframe programmatically
df_olympic[['Age','Height','Weight','Medal']].isnull().sum()

Age         9474
Height     60171
Weight     62875
Medal     231333
dtype: int64

Issue and justification:<br>
Data Completeness: <br>
Olympic data has missing data problems specifically in the columns age, height and weight.<br>
The Medal folder also has a huge chunk of missing value but that is understandable because when athletes don't win any medal they intentionally make it blank. 

### Quality Issue 2:

In [11]:
df_olympic_duplicate = df_olympic[df_olympic.duplicated()]

In [12]:
print(df_olympic_duplicate)

            ID                                 Name Sex   Age  Height  Weight  \
1252       704                   Dsir Antoine Acket   M  27.0     NaN     NaN   
4282      2449               William Truman Aldrich   M  48.0     NaN     NaN   
4283      2449               William Truman Aldrich   M  48.0     NaN     NaN   
4862      2777               Hermann Reinhard Alker   M  43.0     NaN     NaN   
4864      2777               Hermann Reinhard Alker   M  43.0     NaN     NaN   
...        ...                                  ...  ..   ...     ...     ...   
269994  135072  Anna Katrina Zinkeisen (-Heseltine)   F  46.0     NaN     NaN   
269995  135072  Anna Katrina Zinkeisen (-Heseltine)   F  46.0     NaN     NaN   
269997  135072  Anna Katrina Zinkeisen (-Heseltine)   F  46.0     NaN     NaN   
269999  135073   Doris Clare Zinkeisen (-Johnstone)   F  49.0     NaN     NaN   
270200  135173                     Henri Achille Zo   M  58.0     NaN     NaN   

                 Team  NOC 

Issue and justification:<br>
Data Completeness: <br>
Olympic data has 271116 unique values and we can see that there are 1385 duplicated values which should be removed before analysis as it will affect our analysis.

### Tidiness Issue 1:

In [13]:
data_environment.info()
data_environment.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9656 entries, 0 to 9655
Data columns (total 66 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     9656 non-null   int64  
 1   Area          9656 non-null   object 
 2   Months Code   9656 non-null   int64  
 3   Months        9656 non-null   object 
 4   Element Code  9656 non-null   int64  
 5   Element       9656 non-null   object 
 6   Unit          9656 non-null   object 
 7   Y1961         8287 non-null   float64
 8   Y1962         8322 non-null   float64
 9   Y1963         8294 non-null   float64
 10  Y1964         8252 non-null   float64
 11  Y1965         8281 non-null   float64
 12  Y1966         8364 non-null   float64
 13  Y1967         8347 non-null   float64
 14  Y1968         8345 non-null   float64
 15  Y1969         8326 non-null   float64
 16  Y1970         8308 non-null   float64
 17  Y1971         8303 non-null   float64
 18  Y1972         8323 non-null 

Area Code  Months Code  Element Code        Y1961        Y1962  \
count  9656.000000  9656.000000    9656.00000  8287.000000  8322.000000   
mean    821.806338  7009.882353    6674.50000     0.402433     0.315527   
std    1781.072213     6.038255     596.53089     0.701567     0.713777   
min       1.000000  7001.000000    6078.00000    -4.018000    -5.391000   
25%      78.000000  7005.000000    6078.00000     0.057000    -0.033000   
50%     153.500000  7009.000000    6674.50000     0.366000     0.333000   
75%     226.250000  7016.000000    7271.00000     0.676500     0.627000   
max    5873.000000  7020.000000    7271.00000     5.771000     4.373000   

             Y1963        Y1964        Y1965        Y1966        Y1967  ...  \
count  8294.000000  8252.000000  8281.000000  8364.000000  8347.000000  ...   
mean      0.317393     0.269382     0.217839     0.376419     0.263239  ...   
std       0.853133     0.749216     0.739418     0.737370     0.725421  ...   
min      -8.483000    -7.309000    -4.728000    -8.147000    -6.531000  ...   
25%       0.030250    -0.102500    -0.214000     0.055000    -0.169000  ...   
50%       0.355000     0.326000     0.303000     0.360000     0.313000  ...   
75%       0.647750     0.609000     0.584000     0.660250     0.601000  ...   
max       4.666000     5.233000     5.144000     5.771000     4.768000  ...   

             Y2010        Y2011        Y2012        Y2013        Y2014  \
count  8435.000000  8437.000000  8350.000000  8427.000000  8377.000000   
mean      0.884504     0.768488     0.788930     0.829647     0.913872   
std       0.878303     0.750631     0.858586     0.713753     0.815933   
min      -6.072000    -4.854000    -5.785000    -3.642000    -5.367000   
25%       0.392000     0.365000     0.372250     0.408500     0.418000   
50%       0.771000     0.640000     0.651000     0.719000     0.745000   
75%       1.276500     1.091000     1.108000     1.126000     1.190000   
max       7.190000     6.531000    10.826000     6.738000    11.759000   

             Y2015        Y2016        Y2017        Y2018        Y2019  
count  8361.000000  8348.000000  8366.000000  8349.000000  8365.000000  
mean      1.018816     1.081491     1.003342     1.010832     1.094599  
std       0.840189     0.877399     0.809800     0.872199     0.853953  
min      -4.068000    -3.306000    -3.584000    -2.216000    -2.644000  
25%       0.437000     0.457000     0.443000     0.434000     0.455000  
50%       0.858000     0.949000     0.865000     0.810000     0.939000  
75%       1.389000     1.496000     1.364750     1.341000     1.508000  
max       7.590000    10.478000     7.389000     9.228000     7.215000  

[8 rows x 62 columns]

Issue and justification:<br> 
Single observation stored in multiple tables <br>
At first glance of data, we can observe that year entry is stored in multiple columns in environment data, which can be considered as tidiness issue because it is a single observation stored in multiple columns. we will add all years into one entry once we clean this data. 

### Tidiness Issue 2: 

In [14]:
data_environment_duplicate = data_environment[data_environment.duplicated()]
print(data_environment_duplicate)

Empty DataFrame
Columns: [Area Code, Area, Months Code, Months, Element Code, Element, Unit, Y1961, Y1962, Y1963, Y1964, Y1965, Y1966, Y1967, Y1968, Y1969, Y1970, Y1971, Y1972, Y1973, Y1974, Y1975, Y1976, Y1977, Y1978, Y1979, Y1980, Y1981, Y1982, Y1983, Y1984, Y1985, Y1986, Y1987, Y1988, Y1989, Y1990, Y1991, Y1992, Y1993, Y1994, Y1995, Y1996, Y1997, Y1998, Y1999, Y2000, Y2001, Y2002, Y2003, Y2004, Y2005, Y2006, Y2007, Y2008, Y2009, Y2010, Y2011, Y2012, Y2013, Y2014, Y2015, Y2016, Y2017, Y2018, Y2019]
Index: []

[0 rows x 66 columns]


Issue and justification:<br>
Multiple variables are stored in one column.<br>
As we can see from the environmental data, In the element table there is two value stored temperature and standard deviation which is not the best approach as it stores two different values. In the data cleaning process, we will remove this many other tables from the data as they do not help in any statistical analysis.

## 3. Clean data
Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

### Olympic Data Cleaning: 
1. Drop duplicate
    - before removing duplicates 271116 and we removed 1385 duplicate rows and now we have 269731 rows
2. Removing non useful columns
    
### Climate Change:

In [15]:
# Droping duplicates:
df_olympic.shape
df_olympic_clean = df_olympic.drop_duplicates()
#df_olympic.shape
df_olympic_clean.shape
df_olympic_clean.head()
#df_olympic_clean['Sport'].unique()
#df_olympic_clean['Sport'].nunique()

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season       City          Sport  \
0  CHN  1992 Summer  1992  Summer  Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer     London           Judo   
2  DEN  1920 Summer  1920  Summer  Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer      Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter    Calgary  Speed Skating   

                              Event Medal  
0       Basketball Men's Basketball   NaN  
1      Judo Men's Extra-Lightweight   NaN  
2           Football Men's Football   NaN  
3       Tug-Of-War Men's Tug-Of-War  Gold  
4  Speed Skating Women's 500 metres   NaN

In [16]:
# removing non useful columns
df_olympic_clean = df_olympic_clean.drop(columns = ["Height","Weight","NOC"], axis=1)
df_olympic_clean.head()


ID                      Name Sex   Age            Team        Games  Year  \
0   1                 A Dijiang   M  24.0           China  1992 Summer  1992   
1   2                  A Lamusi   M  23.0           China  2012 Summer  2012   
2   3       Gunnar Nielsen Aaby   M  24.0         Denmark  1920 Summer  1920   
3   4      Edgar Lindenau Aabye   M  34.0  Denmark/Sweden  1900 Summer  1900   
4   5  Christine Jacoba Aaftink   F  21.0     Netherlands  1988 Winter  1988   

   Season       City          Sport                             Event Medal  
0  Summer  Barcelona     Basketball       Basketball Men's Basketball   NaN  
1  Summer     London           Judo      Judo Men's Extra-Lightweight   NaN  
2  Summer  Antwerpen       Football           Football Men's Football   NaN  
3  Summer      Paris     Tug-Of-War       Tug-Of-War Men's Tug-Of-War  Gold  
4  Winter    Calgary  Speed Skating  Speed Skating Women's 500 metres   NaN

In [17]:
#df_olympic_clean = df.fillna(value=0)
df_olympic_clean = df_olympic_clean.fillna(value=0)
df_olympic_clean['Age'] = df_olympic_clean['Age'].astype('int')
df_olympic_clean.head(5)



ID                      Name Sex  Age            Team        Games  Year  \
0   1                 A Dijiang   M   24           China  1992 Summer  1992   
1   2                  A Lamusi   M   23           China  2012 Summer  2012   
2   3       Gunnar Nielsen Aaby   M   24         Denmark  1920 Summer  1920   
3   4      Edgar Lindenau Aabye   M   34  Denmark/Sweden  1900 Summer  1900   
4   5  Christine Jacoba Aaftink   F   21     Netherlands  1988 Winter  1988   

   Season       City          Sport                             Event Medal  
0  Summer  Barcelona     Basketball       Basketball Men's Basketball     0  
1  Summer     London           Judo      Judo Men's Extra-Lightweight     0  
2  Summer  Antwerpen       Football           Football Men's Football     0  
3  Summer      Paris     Tug-Of-War       Tug-Of-War Men's Tug-Of-War  Gold  
4  Winter    Calgary  Speed Skating  Speed Skating Women's 500 metres     0

In [18]:
df_olympic_clean['Age'].fillna(df_olympic_clean['Age'].mean(),inplace=True)

#df_olympic_clean.info()


In [19]:
count = df_olympic_clean['Age'].value_counts()[0]
print(count)

9315


### **Quality Issue 1: FILL IN**

In [20]:
# FILL IN - Apply the cleaning strategy

In [21]:
# FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [22]:
#FILL IN - Apply the cleaning strategy

In [23]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [24]:
#FILL IN - Apply the cleaning strategy

In [25]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 2: FILL IN**

In [26]:
#FILL IN - Apply the cleaning strategy

In [27]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [28]:
#FILL IN - Remove unnecessary variables and combine datasets

## 4. Update your data store
Update your local database/data store with the cleaned data, following best practices for storing your cleaned data:

- Must maintain different instances / versions of data (raw and cleaned data)
- Must name the dataset files informatively
- Ensure both the raw and cleaned data is saved to your database/data store

In [29]:
#FILL IN - saving data

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [30]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [31]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN